drivers/azure/azure.go

```diff
@@ -0,0 +1,464 @@
+	//     virtual network is in a different region.
+	//   - Changing IP Address space of a subnet would fail if there are machines
+	//     running in the Virtual Network.
+	log.Info("Completed machine pre-create checks.")
+	return nil
+}
+
+// Create creates the virtual machine.
+func (d *Driver) Create() error {
+	// NOTE(ahmetalpbalkan): We can probably parallelize the sh*t out of this.
+	// However that would lead to a concurrency logic and while creation of a
+	// resource fails, other ones would be kicked off, which could lead to a
+	// resource leak. This is slower but safer.
+	c, err := d.newAzureClient()
+	if err != nil {
+		return err
+	}
+
+	if err := c.CreateResourceGroup(d.ResourceGroup, d.Location); err != nil {
+		return err
+	}
+	if err := c.CreateAvailabilitySetIfNotExists(d.ctx, d.ResourceGroup, d.AvailabilitySet, d.Location); err != nil {
+		return err
+	}
+	if err := c.CreateNetworkSecurityGroup(d.ctx, d.ResourceGroup, d.naming().NSG(), d.Location, d.ctx.FirewallRules); err != nil {
+		return err
+	}
+	if err := c.CreateVirtualNetworkIfNotExists(d.ResourceGroup, d.VirtualNetwork, d.Location); err != nil {
+		return err
+	}
+	if err := c.CreateSubnet(d.ctx, d.ResourceGroup, d.VirtualNetwork, d.SubnetName, d.SubnetPrefix); err != nil {
+		return err
+	}
+	if d.NoPublicIP {
+		log.Info("Not creating a public IP address.")
+	} else {
+		if err := c.CreatePublicIPAddress(d.ctx, d.ResourceGroup, d.naming().IP(), d.Location); err != nil {
+			return err
+		}
+	}
+	if err := c.CreateNetworkInterface(d.ctx, d.ResourceGroup, d.naming().NIC(), d.Location,
+		d.ctx.PublicIPAddressID, d.ctx.SubnetID, d.ctx.NetworkSecurityGroupID, d.PrivateIPAddr); err != nil {
+		return err
+	}
+	if err := c.CreateStorageAccount(d.ctx, d.ResourceGroup, d.Location, defaultStorageType); err != nil {
+		return err
+	}
+	if err := d.generateSSHKey(d.ctx); err != nil {
+		return err
+	}
+	if err := c.CreateVirtualMachine(d.ResourceGroup, d.naming().VM(), d.Location, d.Size, d.ctx.AvailabilitySetID,
+		d.ctx.NetworkInterfaceID, d.BaseDriver.SSHUser, d.ctx.SSHPublicKey, d.Image, d.ctx.StorageAccount); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Remove deletes the virtual machine and resources associated to it.
+func (d *Driver) Remove() error {
+	// NOTE(ahmetalpbalkan):
+	//   - remove attemps are best effort and if a resource is already gone, we
+	//     continue removing other resources instead of failing.
+	//   - we can probably do a lot of parallelization here but a sequential
+	//     logic works fine too. If we were to detach the NIC from the VM and
+	//     then delete the VM, this could enable some parallelization.
+
+	log.Info("NOTICE: Please check Azure portal/CLI to make sure you have no leftover resources to avoid unexpected charges.")
+	c, err := d.newAzureClient()
+	if err != nil {
+		return err
+	}
+	if err := c.DeleteVirtualMachineIfExists(d.ResourceGroup, d.naming().VM()); err != nil {
```

Comment:
@ppadala  No, I don't think so. We create two types of resources:

1. Resources who are attached to VM: (e.g. Network interface, IP address, Firewall) you can't delete those without deleting the VM or detaching those resources first. Deleting VM just disassociates these resources so that Delete calls below work just fine.

2. Resources that hold the VM: (e.g. Resource Group, Availability Set, Virtual Network, Subnet) You can't remove those without removing the VM first, it's like deleting a non-empty folder. The API does not allow it.